v2/failure-injected-spanner-service/src/main/java/com/google/cloud/teleport/v2/spanner/service/SpannerService.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2025 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.spanner.service;
+
+import com.google.cloud.ServiceFactory;
+import com.google.cloud.spanner.Spanner;
+import com.google.cloud.spanner.SpannerOptions;
+import com.google.cloud.spanner.spi.v1.SpannerInterceptorProvider;
+import com.google.cloud.teleport.v2.failureinjection.ErrorInjectionPolicy;
+import com.google.cloud.teleport.v2.failureinjection.ErrorInjectionPolicyFactory;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.io.Serializable;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SpannerService implements ServiceFactory<Spanner, SpannerOptions>, Serializable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SpannerService.class);
+
+  private ErrorInjectionPolicy errorInjectionPolicy;
+
+  /** Injects errors in streaming calls to simulate call restarts. */
+  private static class GrpcErrorInjector implements ClientInterceptor, Serializable {
+
+    private ErrorInjectionPolicy errorInjectionPolicy;
+    private final Random random = new Random();
+
+    GrpcErrorInjector(ErrorInjectionPolicy errorInjectionPolicy) {
+      this.errorInjectionPolicy = errorInjectionPolicy;
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+        final MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+      // Only inject errors in the Cloud Spanner data API.
+      if (!method.getFullMethodName().startsWith("google.spanner.v1.Spanner")) {
+        return next.newCall(method, callOptions);
+      }
+      if (method.getFullMethodName().startsWith("google.spanner.v1.Spanner/BatchCreateSessions")
+          || method.getFullMethodName().startsWith("google.spanner.v1.Spanner/CreateSession")) {
+        return next.newCall(method, callOptions);
+      }
```

Comment:
Nit: It's a bit confusing, you could refactor this out to a separate filter method.

Reply:
Done.